Refactor sysinfo

seven-beep opened the pull request:
> Sorry, but could you rebase your modifications on the main branch?

Here we go :)

original message: 

I am not sure all will interest you but I have passed a few hours to refactor the sysinfo python script for my own usage :

   - use python3
   - dropped utmp for a more trivial output with the command "who"
   -  use the findmnt command to filter the mounts and produce a more compact output
   - added error handling on theses statfs calls
   - restored the ip of the main interface
   - use a main function

seven-beep committed:
Restore ip address

seven-beep committed:
Drop utmp dependency

seven-beep committed:
Use findmnt to parse the filesystem

There is a lot that can go wrong if we ask for bind mounts, containers, etc.
Stay concise.  Also rearrange the output in columns so it is more compact.

seven-beep committed:
Add a main function

seven-beep committed:
Update commentaries

seven-beep committed:
Adjust padding

seven-beep committed:
Simpler ip functions

seven-beep committed:
Update README (and padding a little)

seven-beep committed:
Update headers of sysinfo

seven-beep committed:
sysinfo: Use the ip of the default address

seven-beep committed:
sysinfo: Do not track tmpfs

seven-beep committed:
sysinfo: update formating

ldidry commented on README.md:
````diff
@@ -55,7 +55,7 @@ sudo apt install ./dynamic-motd_*.deb
 
 ```
-apt-get install figlet lsb-release python3-utmp bc
+apt-get install figlet lsb-release bc
````
`findmnt` is in `util-linux`, could you add it, please?

ldidry commented on update-motd.d/sysinfo.py:
```diff
@@ -3,109 +3,166 @@
+                "--noheading",
+                "--real",
+                "--uniq",
+                "--fstab",
```
With `--fstab`, you won’t show disks mounted with [systemd mount units](https://www.freedesktop.org/software/systemd/man/latest/systemd.mount.html).

ldidry commented:
To be honest, I don’t like displaying the IP address in the banner. You choose to display only one IP address into the many possibles (it may not suite all users) and `ip route list dev` won’t work on IPv6-only machines.

And I prefer having multiple login showing, with their IP address and time, as:
- it allows me to know if I’m already logged
- it allows me to see all the IP addresses of those connections (better for security)
- if a co-worker has an old session on the server, I can ask them if it’s on purpose or if they can logout, allowing me, for ex., to reboot the server

We may have a config file to change the behavior of the script, like the `hushlogin` file.

seven-beep commented:
Luc Didry ***@***.***> writes:

> To be honest, I don’t like displaying the IP address in the banner. You choose
> to display only one IP address into the many possibles (it may not suite all
> users) and ip route list dev won’t work on IPv6-only machines.
>
> And I prefer having multiple login showing, with their IP address and time, as:
>
> • it allows me to know if I’m already logged
> • it allows me to see all the IP addresses of those connections (better for
>   security)
> • if a co-worker has an old session on the server, I can ask them if it’s on
>   purpose or if they can logout, allowing me, for ex., to reboot the server
>
> We may have a config file to change the behavior of the script, like the
> hushlogin file.

That is fair, give me some time to address it.

Concerning the findmnt command, I choose the '--fstab' flag because I
want to keep the output terse by eg, not displaying all bind mount that
can be generated by docker or other software.

I can also revise that and/or address it via configurations options.

seven-beep committed:
Restore usage of utmp python module

seven-beep committed:
Remove --fstab from findmnt call in sysinfo

Initially I used this argument as I work with systems with a lot of bind mounts
and it felt unnecessary to list them all, eg:

  Mount points                         Disk usage       Inodes usage
 /                                   49%  of 19.3G     12.8% of 1299984
 /rw                                 35%  of 20.5G     26.3% of 1373568
 /home                               35%  of 20.5G     26.3% of 1373568
 /usr/local                          35%  of 20.5G     26.3% of 1373568
 /var/spool/cron                     35%  of 20.5G     26.3% of 1373568
 /etc/hosts                          35%  of 20.5G     26.3% of 1373568
 /var/lib/systemd/timers             35%  of 20.5G     26.3% of 1373568
 /var/lib/plocate                    35%  of 20.5G     26.3% of 1373568
 /var/lib/containerd                 35%  of 20.5G     26.3% of 1373568
 /etc/docker                         35%  of 20.5G     26.3% of 1373568
 /var/lib/docker                     35%  of 20.5G     26.3% of 1373568
 /home/user/.emacs.d/elpa            2%   of 3.9G      2.4% of 262144
 /rw/home/user/.emacs.d/elpa         2%   of 3.9G      2.4% of 262144
 /proc/xen                           -    of 0         100.0% of 00

However it seems impossible to filter theses mounts in one call,
(https://unix.stackexchange.com/questions/430074/how-to-exclude-bind-mounts-from-findmnt-results-list)
and --fstab is a bad compromise as mounts points may be declared in many others
ways than /etc/fstab.

seven-beep committed:
Remove default device IPv4 detection

seven-beep commented:
Hello @ldidry,

Finally after following your indication, this pull request become mainly
about using `findmnt` to construct a more compact view of the filesystems.

What do you think ?

seven-beep committed:
feat: add a timeout decorator to filesystem function

A filesystem may freeze the script if it is not accessible.

This had for effect to block ssh access in case of an indisponible remote
filesystem.

seven-beep committed:
fix: unnecessary warnings on apt invocation

seven-beep committed:
feat: avoid xenfs in filesystem report

seven-beep committed:
fix upgradable packages must report 0 when 0 are upgradable

apt will output something like 'Listing ... Done' which was counted as a package.

seven-beep committed:
chore: changing reboot required message

/var/run/reboot-required may be touched for other reasons that a pending kernel update.
